sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/ApplicationLifecycleListener.java

```diff
@@ -0,0 +1,148 @@
+     */
+    private Handler mHandler;
+
+    /**
+     * A collection of ApplicationLifecycleCallbacks.
+     */
+    private Collection<ApplicationLifecycleCallbacks> mLifecycleCallbacks = new ArrayList<>();
+
+    private Runnable mDelayedPauseRunnable = new Runnable() {
+        @Override
+        public void run() {
+            dispatchPauseIfNeeded();
+            dispatchStopIfNeeded();
+        }
+    };
+
+    public ApplicationLifecycleListener(Handler handler) {
+        this.mHandler = handler;
+    }
+
+    public void registerApplicationLifecycleCallbacks(ApplicationLifecycleCallbacks lifecycleCallback) {
+        mLifecycleCallbacks.add(lifecycleCallback);
+    }
+
+    private void dispatchPauseIfNeeded() {
+        if (mResumedCounter == 0) {
+            mPauseSent = true;
+        }
+    }
+
+    private void dispatchStopIfNeeded() {
+        if (mStartedCounter == 0 && mPauseSent) {
+            for (ApplicationLifecycleCallbacks service : mLifecycleCallbacks) {
+                service.onApplicationStopped();
+            }
+            mStopSent = true;
+        }
+    }
+
+    @Override
+    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {
+    }
+
+    @Override
+    public void onActivityStarted(@NonNull Activity activity) {
+        mStartedCounter++;
+        if (mStartedCounter == 1 && mStopSent) {
+            for (ApplicationLifecycleCallbacks service : mLifecycleCallbacks) {
+                service.onApplicationStarted();
+            }
+            mStopSent = false;
+        }
+    }
+
+    @Override
+    public void onActivityResumed(@NonNull Activity activity) {
+        mResumedCounter++;
+        if (mResumedCounter == 1) {
+            if (mPauseSent) {
+                mPauseSent = false;
+            } else {
+                mHandler.removeCallbacks(mDelayedPauseRunnable);
+            }
+        }
+    }
+
+    @Override
+    public void onActivityPaused(@NonNull Activity activity) {
+        mResumedCounter--;
+        if (mResumedCounter == 0) {
+
+            /*
+             * OnPause and onStop events will be dispatched with a delay after a last activity
+             * passed through them. This delay is long enough to guarantee that
+             * ApplicationLifecycleListener won't send any events if activities are destroyed
+             * and recreated due to a configuration change.
+             */
+            mHandler.postDelayed(mDelayedPauseRunnable, TIMEOUT_MS);
+        }
+    }
+
+    @Override
+    public void onActivityStopped(@NonNull Activity activity) {
+        mStartedCounter--;
+        dispatchStopIfNeeded();
+    }
+
+    @Override
+    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {
+    }
+
+    @Override
+    public void onActivityDestroyed(@NonNull Activity activity) {
+    }
+
+    public interface ApplicationLifecycleCallbacks {
+        void onApplicationStarted();
```

Comment:
How about we rename to something closer to iOS using the foreground / background terminology.